pkg/server/plugin/upstreamauthority/spire/spire_server_client.go

```diff
@@ -0,0 +1,132 @@
+package spireplugin
+
+import (
+	"context"
+	"crypto/x509"
+	"fmt"
+	"sync"
+
+	"github.com/spiffe/go-spiffe/v2/spiffetls/tlsconfig"
+	"github.com/spiffe/go-spiffe/v2/workloadapi"
+	"github.com/spiffe/spire/pkg/common/x509util"
+	"github.com/spiffe/spire/proto/spire/api/server/bundle/v1"
+	"github.com/spiffe/spire/proto/spire/api/server/svid/v1"
+	"github.com/spiffe/spire/proto/spire/common"
+	"github.com/spiffe/spire/proto/spire/types"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+)
+
+// newServerClient creates a new spire-sever client
+func newServerClient(serverAddr string, workloadapiSocket string) *serverClient {
+	return &serverClient{
+		serverAddr:        serverAddr,
+		workloadAPISocket: workloadapiSocket,
+	}
+}
+
+type serverClient struct {
+	conn              *grpc.ClientConn
+	serverAddr        string
+	workloadAPISocket string
+
+	mtx    sync.RWMutex
+	source *workloadapi.X509Source
+
+	bundleClient bundle.BundleClient
+	svidClient   svid.SVIDClient
+}
+
+// start initializes spire-server endpoints client, it uses X509 source to keep an active connection
+func (c *serverClient) start(ctx context.Context) error {
+	source, err := workloadapi.NewX509Source(ctx, workloadapi.WithClientOptions(workloadapi.WithAddr(c.workloadAPISocket)))
+	if err != nil {
+		return fmt.Errorf("unable to create X509Source: %v", err)
+	}
+
+	tlsConfig := tlsconfig.MTLSClientConfig(source, source, tlsconfig.AuthorizeAny())
```

Comment:
The peer needs to have the SPIRE server SPIFFE ID (i.e. `tlsconfig.AuthorizeID(serverID)`)